tidy3d/components/parameter_perturbation.py

```diff
@@ -1227,6 +1228,389 @@ def _sample_delta_eps_delta_sigma(
+                    7.11e-16,
+                    5.29e-16,
+                    9.72e-16,
+                    1.22e-15,
+                    1.16e-15,
+                    3.16e-15,
+                    1.51e-14,
+                    2.71e-14,
+                    2.65e-14,
+                    2.94e-14,
+                    6.85e-14,
+                    2.60e-13,
+                ],
+            ),
+            "s": (
+                "wvl",
+                [
+                    0.835,
+                    0.838,
+                    0.841,
+                    0.832,
+                    0.849,
+                    0.834,
+                    0.839,
+                    0.838,
+                    0.833,
+                    0.826,
+                    0.812,
+                    0.812,
+                    0.807,
+                    0.776,
+                    0.769,
+                    0.774,
+                    0.783,
+                    0.772,
+                    0.769,
+                    0.772,
+                    0.750,
+                    0.716,
+                    0.704,
+                    0.706,
+                    0.705,
+                    0.686,
+                    0.656,
+                ],
+            ),
+        },
+        coords={"wvl": np.array([1.3, 1.55] + list(np.arange(2, 14.5, 0.5)))},
+    )
+
+    ref_freq: pd.NonNegativeFloat = pd.Field(
+        ...,
+        title="Reference Frequency",
+        description="Reference frequency to evaluate perturbation at (Hz).",
+        units=HERTZ,
+    )
+
+    @pd.root_validator(skip_on_failure=True)
+    def _check_freq_in_range(cls, values):
+        """Check that the given frequency is within validity range.
+        If not, issue a warning.
+        """
+
+        freq = values.get("ref_freq")
+        freq_range = (C_0 / 14, C_0 / 1.3)
+
+        if freq < freq_range[0] or freq > freq_range[1]:
+            log.warning(
+                f"Given frequency {freq} Hz is outside of the validity range ({freq_range[0]} - "
+                f"{freq_range[1]} Hz) of the Nedeljkovic-Soref-Mashanovich model."
+            )
+
+        return values
+
+    @cached_property
+    def ref_wavelength(self) -> float:
+        """Reference wavelength (um) to evaluate perturbation at."""
+        return C_0 / self.ref_freq
+
+    @cached_property
+    def _coeffs_at_ref_freq(self) -> xr.Dataset:
+        """Coefficients at reference frequency."""
+        return self.perturb_coeffs.interp(wvl=self.ref_wavelength)
+
+    def delta_k(self) -> ChargePerturbationType:
+        """Return the perturbation range of the model."""
+
+        # create free-carrier mesh
+        # NOTE: this range (0, 1e22) should work in most cases. If need be we can add a new field
+        # to the model to allow for custom ranges.
+        Ne_range = np.concatenate(([0], np.logspace(-6, 22, num=200)))
+        Nh_range = np.concatenate(([0], np.logspace(-6, 22, num=200)))
```

Comment:
yeah, I would say lets go ahead and add two new fields in this class `electron_concentration_grid = np.logspace(-6, 22, num=200)` and `hole_concentration_grid = np.logspace(-6, 22, num=200)`